libraries/chain/wasm_interface.cpp

```diff
@@ -228,19 +226,22 @@ namespace eosio { namespace chain {
                   instance = instantiateModule(*module, std::move(link_result.resolvedImports));
                   FC_ASSERT(instance != nullptr);
 
-                  MemoryInstance* current_memory = Runtime::getDefaultMemory(instance);
-
-                  if(current_memory) {
-                     char *mem_ptr = &memoryRef<char>(current_memory, 0);
-                     const auto allocated_memory = Runtime::getDefaultMemorySize(instance);
-                     for (uint64_t i = 0; i < allocated_memory; ++i) {
-                        if (mem_ptr[i])
-                           mem_end = i + 1;
-                     }
-                     mem_image.resize(mem_end);
-                     memcpy(mem_image.data(), mem_ptr, mem_end);
+                  //populate the module's data segments in to a vector so the initial state can be
+                  // restored on each invocation
+                  //Be Warned, this may need to be revisited when module imports make sense. The
+                  // code won't handle data segments that initalize an imported memory which I think
+                  // is valid.
+                  for(const DataSegment& data_segment : module->dataSegments) {
```

Comment:
If there is anything that could use some extra eyeballs, it's this area. Mainly want to double check no off by one.